### PR TITLE
Fix TimerAwaitable rooting

### PR DESCRIPTION
**Pull request by davidfowl**
- This fixes an issue where components that use timer awaitable currently need to be explicitly stopped and disposed for it to be unrooted. This makes it possible to write a finalizer.

This pattern exists in a couple more places in the BCL:
- https://github.com/dotnet/runtime/blob/73b5ef88291d85b6c9393a5829fce25fd939aed3/src/libraries/Common/src/System/Net/WebSockets/ManagedWebSocket.cs#L209-L212
- https://github.com/dotnet/runtime/blob/73b5ef88291d85b6c9393a5829fce25fd939aed3/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs#L108-L112

It usually affects client APIs where disposal of resources happen if something is unreferenced.

### CODE REVIEW

File: src/SignalR/clients/csharp/Client/test/UnitTests/TimerAwaitableTests.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        public void FinalizerRunsIfTimerAwaitableReferencesObject()
+        {
+            var tcs = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+            UseTimerAwaitableAndUnref(tcs);
+
+            // Make sure it *really* cleans up
+            for (int i = 0; i < 5; i++)
+            {
+                GC.Collect();
+                GC.WaitForPendingFinalizers();
```
Comment (anurse) [inline src/SignalR/clients/csharp/Client/test/UnitTests/TimerAwaitableTests.cs:21]:
```suggestion
                GC.Collect();
                GC.WaitForPendingFinalizers();
                if (tcs.Task.IsCompleted)
                {
                	return;
                }
```

That way we can exit early if the task finishes early.

File: src/SignalR/clients/csharp/Client/test/UnitTests/TimerAwaitableTests.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
+using System;
```
Comment (BrennanConroy) [inline src/SignalR/clients/csharp/Client/test/UnitTests/TimerAwaitableTests.cs:1]: copyright

File: src/SignalR/clients/csharp/Client/test/UnitTests/TimerAwaitableTests.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void UseTimerAwaitableAndUnref(TaskCompletionSource<object> tcs)
+        {
+            _ = new ObjectWithTimerAwaitable(tcs).Start();
+        }
+    }
+
+    // This object holds onto a TimerAwaitable referencing the callback (the async continuation is the callback)
+    // it also has a finalizer that triggers a tcs to callers can be notified when this object is being cleaned up.
```
Comment (BrennanConroy) [inline src/SignalR/clients/csharp/Client/test/UnitTests/TimerAwaitableTests.cs:36]:
```suggestion
    // it also has a finalizer that triggers a tcs so callers can be notified when this object is being cleaned up.
```

**Review by BrennanConroy — APPROVED** [review]

**Pull request merged**